src/hooks/useTransactions.tsx

```diff
@@ -0,0 +1,95 @@
+import React from "react";
+import { ethers } from "ethers";
+
+type Transaction = ethers.providers.TransactionResponse;
+type TransactionReceipt = ethers.providers.TransactionRequest;
+type TransactionStatus = "idle" | "pending" | "resolved" | "rejected";
+
+type TransactionState = {
+  transaction?: Transaction;
+  status?: TransactionStatus;
+  error?: Error;
+};
+
+enum ActionType {
+  ERROR,
+  UPDATE,
+}
+
+type Action =
+  | {
+      type: ActionType.UPDATE;
+      payload: TransactionState;
+    }
+  | {
+      type: ActionType.ERROR;
+      payload: Pick<TransactionState, "error">;
+    };
+function reducer(state: TransactionState, action: Action) {
+  switch (action.type) {
+    case ActionType.UPDATE: {
+      return {
+        ...state,
+        ...action.payload,
+      };
+    }
+    case ActionType.ERROR: {
+      const { error } = action.payload;
+      return {
+        ...state,
+        error,
+      };
+    }
+  }
+}
+
+type TransactionsManagerState = {
+  addTransaction: (
+    tx: Transaction,
+    onComplete?: (tx: TransactionReceipt) => void
+  ) => void;
+} & TransactionState;
+function useTransactionsManager(): TransactionsManagerState {
+  const [state, dispatch] = React.useReducer(reducer, {});
+  const { transaction, status, error } = state;
+  const addTransaction = React.useCallback(
+    (tx: Transaction, onComplete?: (tx: TransactionReceipt) => void) => {
+      dispatch({
+        type: ActionType.UPDATE,
+        payload: { transaction: tx, status: "pending" },
+      });
+      tx.wait().then((txReceipt) => {
```

Comment:
i think this manager is overly complex since you only every resolve the a single transaction through a promise. it seems like all you need is a global state so you can communicate there is a transaction in process across components, and then you could just wrap the transaction locally with a react query when submitting it.